src/Mvc/Mvc.Analyzers/src/Startup/StartupAnalysisContext.cs

```diff
@@ -1,22 +1,98 @@
 ﻿// Copyright (c) .NET Foundation. All rights reserved.
 // Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
 
-using Microsoft.CodeAnalysis.Diagnostics;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using Microsoft.CodeAnalysis;
 
 namespace Microsoft.AspNetCore.Analyzers
 {
-    internal class StartupAnalysisContext
+    internal class StartupAnalyzerContext
     {
-#pragma warning disable RS1012 // Start action has no registered actions.
-        public StartupAnalysisContext(OperationBlockStartAnalysisContext operationBlockStartAnalysisContext, StartupSymbols startupSymbols)
+        private readonly object _lock;
+        private readonly Dictionary<INamedTypeSymbol, List<object>> _analysesByType;
+        private readonly StartupAnalzyer _analyzer;
+
+        public StartupAnalyzerContext(StartupAnalzyer analyzer, StartupSymbols startupSymbols)
         {
-            OperationBlockStartAnalysisContext = operationBlockStartAnalysisContext;
+            _analyzer = analyzer;
             StartupSymbols = startupSymbols;
-        }
 
-        public OperationBlockStartAnalysisContext OperationBlockStartAnalysisContext { get; }
+            _analysesByType = new Dictionary<INamedTypeSymbol, List<object>>();
+            _lock = new object();
+        }
 
         public StartupSymbols StartupSymbols { get; }
-#pragma warning restore RS1012 // Start action has no registered actions.
+
+        public T? GetRelatedSingletonAnalysis<T>(INamedTypeSymbol type) where T : class
+        {
+            lock (_lock)
+            {
+                if (_analysesByType.TryGetValue(type, out var list))
```

Comment:
Would it benefit if the value was an `ImmutableList` \ `ImmutableArray` and then do the iteration outside the `lock` or just use a `ConcurrentDictionary` at that point?

Reply:
I have no reason to believe the lock is causing a perf problem. Ultimately it makes sense to split this context class into a builder + results for design clarity though. I'll try it out.